chainstate/src/detail/spend_cache/mod.rs

```diff
@@ -498,6 +602,191 @@ impl<'a, S: BlockchainStorageRead> CachedInputs<'a, S> {
         Ok(())
     }
 
+    fn check_connected_transfer_data(
+        &self,
+        block_id: Id<Block>,
+        tx: &Transaction,
+        token_id: &TokenId,
+        amount: &Amount,
+    ) -> Result<(), TokensError> {
+        // Collect token inputs
+        let total_value_tokens = self
+            .calculate_assets_total_inputs(tx.inputs())
+            .map_err(|_err| TokensError::NoTokenInInputs(tx.get_id(), block_id))?;
+
+        // Is token exist in inputs?
+        let (_, origin_amount) = total_value_tokens
+            .iter()
+            .find(|&(origin_token_id, _)| origin_token_id == token_id)
+            .ok_or_else(|| TokensError::NoTokenInInputs(tx.get_id(), block_id))?;
+
+        // Check amount
+        ensure!(
+            origin_amount >= amount,
+            TokensError::InsuffienceTokenValueInInputs(tx.get_id(), block_id,)
+        );
+
+        Ok(())
+    }
+
+    fn check_connected_issue_data(
```

Comment:
So the token can be infinitely long?

Please check the rules that we used in substrate and implement them here.

Reply:
These checks are done in `check_block`.